lib/job-runner/src/main.rs

```diff
@@ -0,0 +1,243 @@
+
+	// Catch SIGTERM and forward to child
+	//
+	// This will wait for the child to exit and then exit itself so we have time to ship all of the
+	// required logs
+	let mut signals = Signals::new(&[SIGTERM])?;
+	let runc_container_id = container_id.clone();
+	thread::spawn(move || {
+		for _ in signals.forever() {
+			println!("Received SIGTERM, forwarding to runc container {runc_container_id}");
+			let status = Command::new("runc")
+				.arg("kill")
+				.arg("--all")
+				.arg(&runc_container_id)
+				.arg("SIGTERM")
+				.status();
+			println!("runc kill status: {:?}", status);
+			break;
+		}
+	});
+
+	// Ship stdout & stderr logs
+	let stdout_handle = ship_logs(msg_tx.clone(), log_shipper::StreamType::StdOut, runc_stdout);
+	let stderr_handle = ship_logs(msg_tx.clone(), log_shipper::StreamType::StdErr, runc_stderr);
+
+	// Wait for threads to finish
+	match stdout_handle.join() {
+		Result::Ok(_) => {}
+		Err(err) => {
+			eprintln!("stdout thread panicked: {err:?}")
+		}
+	}
+	match stderr_handle.join() {
+		Result::Ok(_) => {}
+		Err(err) => {
+			eprintln!("stderr thread panicked: {err:?}")
+		}
+	}
+
+	// Wait for runc to exit
+	let runc_exit_code = match runc_child.wait() {
+		Result::Ok(exit_code) => {
+			if let Some(exit_code) = exit_code.code() {
+				println!("runc exited with code {}", exit_code);
+				exit_code
+			} else {
+				eprintln!("Unable to read exit code");
+				1
+			}
+		}
+		Err(err) => {
+			eprintln!("Failed to wait for runc: {err:?}");
+			1
+		}
+	};
+
+	// Shutdown
+	match shutdown_tx.send(()) {
+		Result::Ok(_) => {
+			println!("Sent shutdown signal");
+		}
+		Err(err) => {
+			eprintln!("Failed to send shutdown signal: {err:?}");
+		}
+	}
+
+	std::process::exit(runc_exit_code)
+}
+
+// TODO: Rate limit
+
+/// Spawn a thread to ship logs from a stream to log_shipper::LogShipper
+fn ship_logs(
+	msg_tx: mpsc::SyncSender<log_shipper::ReceivedMessage>,
+	stream_type: log_shipper::StreamType,
+	stream: impl BufRead + Send + 'static,
+) -> thread::JoinHandle<()> {
+	std::thread::spawn(move || {
+		// Reduces logging spikes. This logging is in place in order to ensure that a single
+		// spike of logs does not exhaust the long rate limit.
+		//
+		// 64 logs/s
+		let mut throttle_short = throttle::Throttle::new(960, Duration::from_secs(15));
+
+		// Reduces logs from noisy games. Set reasonable caps on how
+		// much can be logged per minute. This is here to prevent games
+		// that log as fast as possible (i.e. positions of objects every
+		// tick) from exhausting the system while still allowing sane
+		// amounts of logging. This happens very frequently.
+		//
+		// 4 logs/s * 1024 bytes/log = 4096 bytes/lobby/s = 14.7 MB/lobby/hr = 353.8 MB/lobby/day  = 10.6 GB/lobby/month
+		let mut throttle_long = throttle::Throttle::new(1200, Duration::from_secs(300));
+
+		// Throttles error logs
+		let mut throttle_error = throttle::Throttle::new(1, Duration::from_secs(60));
```

Comment:
I feel like this throttle might too much; What if for some reason its an error with a stack trace gets printed to stderr, would be nice to get the full stack trace.

Maybe something like 50 every 10 minutes?

Reply:
`throttle_error` is different than stderr, this is only for limiting how frequently we log the rate limit error internal. Both stdout and stderr have the same throttle.

Reply:
Ah okay, I see now, makes sense